instrumentation/java-concurrent/src/main/java/io/opentelemetry/instrumentation/auto/javaconcurrent/KotlinProbeInstrumentation.java

```diff
@@ -0,0 +1,196 @@
+    return named("kotlin.coroutines.jvm.internal.DebugProbesKt");
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    final Map<ElementMatcher<? super MethodDescription>, String> transformers = new HashMap<>();
+    transformers.put(
+        named("probeCoroutineCreated").and(takesArguments(1)),
+        CoroutineCreatedAdvice.class.getName());
+    transformers.put(
+        named("probeCoroutineResumed").and(takesArguments(1)),
+        CoroutineResumedAdvice.class.getName());
+    transformers.put(
+        named("probeCoroutineSuspended").and(takesArguments(1)),
+        CoroutineSuspendedAdvice.class.getName());
+    return transformers;
+  }
+
+  @Override
+  public String[] helperClassNames() {
+    return new String[] {
+      "io.opentelemetry.instrumentation.auto.javaconcurrent.KotlinProbeInstrumentation$CoroutineWrapper",
+      "io.opentelemetry.instrumentation.auto.javaconcurrent.KotlinProbeInstrumentation$TraceScopeKey",
+      "io.opentelemetry.instrumentation.auto.javaconcurrent.KotlinProbeInstrumentation$CoroutineContextWrapper",
+    };
+  }
+
+  public static class CoroutineCreatedAdvice {
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void exit(
+        @Advice.Return(readOnly = false) kotlin.coroutines.Continuation retVal) {
+      if (!(retVal instanceof CoroutineWrapper)) {
+        retVal = new CoroutineWrapper(retVal);
+      }
+    }
+  }
+
+  public static class CoroutineResumedAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void enter(
+        @Advice.Argument(0) final kotlin.coroutines.Continuation continuation) {
+      continuation.getContext().get(TraceScopeKey.INSTANCE);
+    }
+  }
+
+  public static class CoroutineSuspendedAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void enter(
+        @Advice.Argument(0) final kotlin.coroutines.Continuation continuation) {
+      continuation.getContext().minusKey(TraceScopeKey.INSTANCE);
+    }
+  }
+
+  public static class TraceScopeKey implements CoroutineContext.Key {
+    public static final TraceScopeKey INSTANCE = new TraceScopeKey();
+  }
+
+  public static class CoroutineWrapper implements kotlin.coroutines.Continuation {
+    private final Continuation proxy;
+    private final CoroutineContextWrapper contextWrapper;
+
+    public CoroutineWrapper(Continuation proxy) {
+      this.proxy = proxy;
+      this.contextWrapper = new CoroutineContextWrapper(proxy.getContext());
+    }
+
+    public String toString() {
+      return proxy.toString();
+    }
+
+    @NotNull
+    @Override
+    public CoroutineContext getContext() {
+      return contextWrapper;
+    }
+
+    @Override
+    public void resumeWith(@NotNull Object o) {
+      proxy.resumeWith(o);
+    }
+  }
+
+  public static class CoroutineContextWrapper implements CoroutineContext {
```

Comment:
Is it possible to use `ThreadContextElement` instead of hooking into the lifecycle ourselves? If not, can we add some docs on why that doesn't work? I haven't used kotlin much TBH, but my understanding is that's the standard way of syncing a threadlocal with coroutines, which is what we want to achieve here.

https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-thread-context-element/

Reply:
Also, just brainstorming, but is there an approach that wraps dispatchers using something like https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.html#currentContextExecutor-java.util.concurrent.Executor- ? I have no idea the pros/cons for kotlin :) But in Java, my goto for context propagation is usually a wrapped `Executor` so it just came to mind. I've used [this approach](https://github.com/line/armeria/blob/master/examples/context-propagation/kotlin/src/main/kotlin/example/armeria/contextpropagation/kotlin/MainService.kt#L43) in some of the only kotlin I've ever written outside of build scripts.

Reply:
So, yes, I have experimented with more "public" ways of doing this.  I had a version with the `ThreadContextElement` approach and it passed the tests, but it exposes the attached CoroutineContext in a way that is visible (and modifiable) to the app (in addition to being a bit more awkward to implement owing to the larger implementation surface of `CoroutineContext.Element` and writing this in Java rather than Kotlin).  I got worried about app code clearing or resetting context and accidentally dropping our bits.  I am not a Kotlin expert at all though, so perhaps there's a cleaner/more elegant way to use the `ThreadContextElement` approach that I'm not seeing.

Reply:
FWIW, if it helps, I don't see a reason to avoid writing the instrumentation in kotlin for kotlin if it helps - main overhead of kotlin is usually the size of the runtime which is no an issue when the instrumentation already requires it.

But given it seems we can expect relative stability from the debug probes API, I wouldn't worry too much easier it's a point in case anything was obvious. We can always investigate more if something breaks in the future instead of now.